qrproofs/qnr.go

```diff
@@ -0,0 +1,219 @@
+package qrproofs
+
+import (
+	"errors"
+	"github.com/xlab-si/emmy/common"
+	"github.com/xlab-si/emmy/dlog"
+	"math/big"
+)
+
+// Statistical zero-knowledge proof	of quadratic non-residousity (implemented for historical
+// reasons).
+
+func QNR(y *big.Int, qr *dlog.QR) (bool, error) {
+	prover := NewQNRProver(qr, y)
+	verifier := NewQNRVerifier(qr, y)
+	m := qr.N.BitLen()
+
+	for i := 0; i < m; i++ {
+		w, pairs := verifier.GetChallenge()
+		prover.SetProofRandomData(w)
+		// get challenge from prover for proving that verifier is not cheating
+		randVector := prover.GetChallenge()
+
+		verProof := verifier.GetProofData(randVector)
+
+		verifierIsHonest := prover.Verify(pairs, verProof)
+		if !verifierIsHonest {
+			err := errors.New("verifier is not honest")
+			return false, err
+		}
+
+		typ, err := prover.GetProofData(w)
+		if err != nil {
+			return false, nil
+		}
+
+		proved := verifier.Verify(typ)
+		if proved == false {
+			return false, nil
+		}
+	}
+	return true, nil
+}
+
+type QNRProver struct {
+	QR *dlog.QR
+	Y  *big.Int
+	w  *big.Int
+}
+
+func NewQNRProver(qr *dlog.QR, y *big.Int) *QNRProver {
+	prover := QNRProver{
+		QR: qr,
+		Y:  y,
+	}
+
+	return &prover
+}
+
+func (prover *QNRProver) GetChallenge() []int {
+	m := prover.QR.N.BitLen()
+	var randVector []int
+	for i := 0; i < m; i++ {
+		// todo: remove big.Int
+		b := common.GetRandomInt(big.NewInt(2)) // 0 or 1
+		var r int
+		if b.Cmp(big.NewInt(0)) == 0 {
+			r = 0
+		} else {
+			r = 1
+		}
+		randVector = append(randVector, r)
+	}
+	return randVector
+}
+
+func (prover *QNRProver) SetProofRandomData(w *big.Int) {
+	prover.w = w
+}
+
+func (prover *QNRProver) GetProofData(challenge *big.Int) (int, error) {
+	isQR, err := prover.QR.IsQR(challenge)
+	if err != nil {
+		return 0, err
+	}
+	var typ int
```

Comment:
Can you please rename this to something else :slightly_smiling_face:  (for instance, the same name as in the `QRVerifier`)